### ISSUE TITLE
client.runServerPlugin does not correctly start the plugin

### ISSUE THREAD

**Issue by nawgszy** [webgme/webgme-engine#99]
I am having an extremely bizarre issue. I am running a plugin via client.runServerPlugin, and for every plugin I have created, it works the first time I try to run it. For most of them, they work repeatedly, and running one before the other does not matter. However, when I run a specific plugin, the

    client.runServerPlugin(pluginId, client.getCurrentPluginContext(), callback)

will never see the callback get invoked.

Moreover, on the server side, all my plugins will log out when their `constructor` is called; this is never happening. I haven't been able to trace the actual source of failure, but I know that `src/client/pluginmanager` is correctly invoking `runServerPlugin` and calling `storage.simpleRequest`; here, too, the callback from `storage.simpleRequest` is never invoked, and I cannot tell what the network requests made to the websocket are in the browser's Dev Tools.

This is a particularly poor question, but are there any places I can try and debug this, or any known ways to trigger this behaviour? I tried to listen to the `pluginmanager`'s `dispatchEvent` calls, but those never happen. Really, I have nowhere to start.

**Comment by pmeijer** [comment]
Initially I was thinking that there could be an issue with "hanging" promise rejections. Which can be avoided by ensuring [promise.done](https://github.com/kriskowal/q/wiki/API-Reference#promisedoneonfulfilled-onrejected-onprogress) is called. Which actually could be the case, can you checked the endpoint `/api/status` and check if there running workers?
Things that could help:
1. Set the log-level to debug. Either add a [transport](https://github.com/webgme/webgme-engine/blob/master/config/config.default.js#L167). Or if you use the default you can set the console to debug by: `config.server.log.transports[0].options.level = 'debug'. 
2. Ideally see if you can debug the server and server worker. When running in debug mode the webgme server will log the port for the worker which will receive the next request. That port is where you need to attach a debugger to. (In case you're using webstorm the latter is done via Run/Edit Config../+/"Attach to Node.js/Chrome", which varies with versions though). If it never reaches your plugin, you could put a break-point [here](https://github.com/webgme/webgme-engine/blob/master/src/server/worker/simpleworker.js#L73).

**Comment by nawgszy** [comment]
Ok thank you for the suggestions, here are my results to this point:

I've managed to consistently reproduce the problem, at least. I may run an individual plugin 1 to many times in a row; if, after running any plugin, I try to run any other plugin, it will fail.

As far as I can see, the worker receives the 'initialize' command correctly on this next plugin run, and then when `CONSTANTS.workerCommands.executePlugin` comes through, the `simpleworkermanager` correctly sends the executePlugin command AND spawns a new, `free` state worker

Next, the executePlugin command just hangs forever. Why?

I traced the `src/plugin/managerbase.js` file to see where it fails. I see that in the `src/plugin/workerrequests` file's function `executePlugin`, we successfully get to `pluginManager.executePlugin`. This works up until this code block, inside `getPluginInstance`:

    requirejs([pluginPath], instantiatePlugin,
      function (err) {
        console.log('requirejs failed', err);
        deferred.reject(err);
      }
    );

I have logs inside `instantiatePlugin` function AND what looks like the error handler, and neither of them get invoked. `requirejs` is a bit of a black hole in terms of what I know, any idea how I could corrupt the requirejs config so badly that this line doesn't seem to invoke either of the callbacks it was given?

**Comment by nawgszy** [comment]
I even went a step further and wrapped the requirejs call in a `try { requirejs(...) } catch (e) { console.error(e) }` block and it still failed... So I suppose this answers how there was never any returned error from this. I will see if I can solve what happens to requirejs in the meantime.

**Comment by pmeijer** [comment]
I'm trying to reproduce it here but am so far unsuccessful. A couple of things just to make sure I see your issue correctly (I'm pretty sure about the answer but just want to rule it out for the sake of it).

1. Is there any error logged at all?
2. Does (what seems to be) the causing plugin ever run successfully?

So there is some strange thing with requirejs when trying to load a non-existing module twice (the second time no error is thrown and it simply returns `undefined`). However that only seems to apply when loading a module synchronously (which is not the case here). Further if it would return `undefined` the error from instantiatePlugin at least kills the worker. Even further it should only require the plugin once.. 
Some thing that could potentially lead to it being able to load the plugin some times but not other are if the directories where the plugins live no longer exists or are accessible..
I cannot give you any answer at the moment, located some things where the error handling could be better on our side - but will have to look closer myself at requirejs.

I'm trying to replicate this on windows.. Maybe this is platform dependent. What os are you using?

**Comment by nawgszy** [comment]
I have solved it! It was my fault, I simply abused requirejs a bit and it didn't like that to the point of failing silently. 

I used relative paths in order to load one file from multiple plugins, and as http://requirejs.org/docs/errors.html seems to suggest under 'Load Timeout for Modules', that can lead to a situation where it has multiple IDs for one file and doesn't like that.

I updated my config file to set `config.requirejsPaths.Library = <folder containing files I need>`, and loaded them correctly via `define(['Library/file'], function(fileContents) {...})` and everything now works correctly.

Thanks a bunch for the help, sorry for spending your time on this!
